framework/auth/views.py

```diff
@@ -199,87 +199,109 @@ def forgot_password_post(auth, **kwargs):
-        if not (next_url[0] == '/'
-                or next_url.startswith(settings.DOMAIN)
-                or next_url.startswith(settings.CAS_SERVER_URL)
-                or next_url.startswith(settings.MFR_SERVER_URL)):
-            raise HTTPError(http.InvalidURL)
+    return data, http.OK
 
-    if auth.logged_in:
-        if not log_out:
-            if next_url:
-                return redirect(next_url)
-            return redirect('dashboard')
-        # redirect user to CAS for logout, return here w/o authentication
-        return auth_logout(redirect_url=request.url)
 
-    status_message = request.args.get('status', '')
-    if status_message == 'expired':
-        status.push_status_message('The private link you used is expired.', trust=False)
-        status.push_status_message('The private link you used is expired.  Please <a href="/settings/account/">'
-                                   'resend email.</a>', trust=False)
+@collect_auth
+def auth_login(auth, **kwargs):
+    """
+    This view serves as the entry point for OSF login and campaign login.
+    HTTP Method: GET
 
-    if next_url and must_login_warning:
-        status.push_status_message(language.MUST_LOGIN, trust=False)
+    :param kwargs: campaign, currently `institution` and `prereg`
+    :param kwargs: next, provided as default CAS service url
 
-    # set login_url to form action, upon successful authentication specifically w/o logout=True,
-    # allows for next to be followed or a redirect to the dashboard.
-    redirect_url = web_url_for('auth_login', next=next_url, _absolute=True)
+    1. `GET /login` without campaign parameter
+            redirect to `/dashboard`, CAS will handle whether user is logged in or not
+    2. `GET /login?campaign=prereg`:
+            if user is logged in, redirect to `/prereg`
+            else, redirect to `/register?campaign=prereg` for user to create account or log in
+    3. `GET /login?campaign=institution`:
+            redirect to CAS institution login with `/dashboard` as target service
+    """
 
-    data = {}
-    if campaign and campaign in campaigns.CAMPAIGNS:
-        if (campaign == 'institution' and settings.ENABLE_INSTITUTIONS) or campaign != 'institution':
-            data['campaign'] = campaign
-    data['login_url'] = cas.get_login_url(redirect_url)
-    data['institution_redirect'] = cas.get_institution_target(redirect_url)
-    data['redirect_url'] = next_url
-    data['sign_up'] = request.args.get('sign_up', False)
-    data['existing_user'] = request.args.get('existing_user', None)
+    campaign = request.args.get('campaign')
 
-    return data, http.OK
+    # `GET /login` without campaign, ignore other request parameters
+    if not campaign:
+        return redirect(web_url_for('dashboard'))
+
+    if campaign in campaigns.CAMPAIGNS:
+        # `Get /login?campaign=prereg`
+        if campaign != 'institution':
```

Comment:
Agreed. I was not able to find a better way for preserving old routes other than using this piece of duplicated logic. Now I realize that this can be dried up when we do it the "SPA-way" you mentioned above.